# Analyzing Spreadsheet Data with Python

In this project, we'll learn introduction to the tools that we can use to load spreadsheet data into Python, manipulate it, and write it back to the original file.

## Learning

After completing this guided project, we will be able to:

1.  Understand the definition of a dataset.
2.  Understand the concept of Cell and Sheets in spreadsheet.
3.  Load a sheet in an `xlsx` or `xls` file to Python.
4.  Perform basic spreadsheet operations such as subsetting, filtering, calculating column mean, median, max and mean, etc.
5.  Write to a sheet in an `xlsx` or `xls` file from Python.


## Exercise 1: Let's Define Data!

So, let's talk about data. We've been throwing this term around a lot: analyzing *data*, manipulating *data* and writing *data*. What exactly is this data?

Data is defined as "a collection of numbers, characters, images or other items that provide information of something" (SDM, 3rd CE, Deveaux et al.). In statistical sciences, we store data in files that records *cases* and *variables*. Cases are often deleted due to missing data.

## Exercise 2: Explore a Spreadsheet!

Now that we know the definition of data, we can dive into a spreadsheet!

A spreadsheet is a computer application for organization, analysis, and storage of data in tabular form. `xlsx` and `xls` are the two most popular formats for spreadsheet files. They can be created, opened, and saved using spreadsheet applications such as Microsoft Office Excel and Google Sheets.An `xlsx` or `xls` file consists of one or more sheets inside. The sheets inside are listed at the bottom left of the Excel window.





## Exercise 3: Load Data to Python!

Often, we store the data collected in files. `xlsx` and `xls` are two most common formats. To analyze data in the spreadsheet, the first thing we need to do is to load data tables from spreadsheet files to Python.

For this project, we will use a dataset to work with. You can click [here](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/analysing-spreadsheet-data-with-python/Airport_Data.xlsx) to download the Excel file if you want to open it locally. This Excel file contains some information about airports in Alaska released by the Federal Aviation Association (FAA).

[Pandas](https://pandas.pydata.org/docs/index.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkQuickLabsanalysingspreadsheetdatawithpython28639550-2022-01-01) is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language. It provides convenient ways to work with data tables in Python which we will be utilizing in this guided project.


First, let's install `openpyxl`. This library will come in handy after we analyze our data as we'll be able to export the results into an Excel file.


In [16]:
pip install openpyxl

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [17]:
!pip install pandas



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [13]:
import requests
import os

if not os.path.isfile(r"E:\Python\Python\Analyzing Spreadsheet with Python\Airport_Data.xlsx"):
    r = requests.get("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/analysing-spreadsheet-data-with-python/Airport_Data.xlsx")
    f = open(r"E:\Python\Python\Analyzing Spreadsheet with Python\Airport_Data.xlsx", mode="wb+")
    f.write(r.content)
    f.close()


To load the spreadsheet into Python, we can use pandas's `read_excel()` function. When calling this function, we'll need to specify:

*   The path to the `xls` or `xlsx` file, and
*   The name of the sheet in which you wish to load into Python.

Both will be passed in format of the string.


In [19]:
import pandas as pd

file_path = r"E:\Python\Python\Analyzing Spreadsheet with Python\Airport_Data.xlsx"

# Read the Excel file
df_facilities = pd.read_excel(file_path)


In [20]:
df_facilities

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
0      50009.*A        AIRPORT       'ADK     3/30/2017    AAL           NONE   
1     50016.1*A        AIRPORT       'AKK     3/30/2017    AAL           NONE   
2      50017.*A        AIRPORT       'Z13     3/30/2017    AAL           NONE   
3     50017.1*C  SEAPLANE BASE       'KKI     3/30/2017    AAL           NONE   
4      50020.*A        AIRPORT       'AKI     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
746  50920.12*C  SEAPLANE BASE       '2Y3     3/30/2017    AAL           NONE   
747    50920.*A        AIRPORT       'YAK     3/30/2017    AAL           NONE   
748   50925.1*A        AIRPORT       'A77     3/30/2017    AAL           NONE   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   
750    50990.*A        AIRPORT       'L20     3/30/2017    AAL           NONE   

    State StateName             County CountyState  ... AirportPositionSource  \
0      AK    ALASKA     ALEUTIANS WEST          AK  ...      3RD PARTY SURVEY   
1      AK    ALASKA      KODIAK ISLAND          AK  ...                  NACO   
2      AK    ALASKA             BETHEL          AK  ...                 STATE   
3      AK    ALASKA             BETHEL          AK  ...                   NaN   
4      AK    ALASKA             BETHEL          AK  ...                 STATE   
..    ...       ...                ...         ...  ...                   ...   
746    AK    ALASKA    SKAGWAY-YAKUTAT          AK  ...               FAA-EST   
747    AK    ALASKA    SKAGWAY-YAKUTAT          AK  ...      3RD PARTY SURVEY   
748    AK    ALASKA          KUSKOKWIM          AK  ...                   NaN   
749    AK    ALASKA  KETCHIKAN GATEWAY          AK  ...                   NaN   
750    AK    ALASKA        UPPER YUKON          AK  ...               FAA-EST   

    AirportPositionSourceDate AirportElevationSource  \
0                    00:00:00       3RD PARTY SURVEY   
1                    00:00:00                   NACO   
2                    00:00:00                  STATE   
3                         NaN                    NaN   
4                    00:00:00                  STATE   
..                        ...                    ...   
746                  00:00:00                    NaN   
747                  00:00:00       3RD PARTY SURVEY   
748                       NaN                    NaN   
749                       NaN                    NaN   
750                  00:00:00                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
0                     00:00:00                   NaN              HGR   
1                     00:00:00                   NaN              NaN   
2                     00:00:00                   NaN              NaN   
3                          NaN                   NaN              NaN   
4                     00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
746                        NaN                   NaN              TIE   
747                   00:00:00                   NaN              HGR   
748                        NaN                   NaN              NaN   
749                        NaN                   NaN              NaN   
750                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
0           CARGO             Y           PADK           SS-SR  
1             NaN             Y           PAKH             NaN  
2           CARGO           Y-L            NaN         SEE RMK  
3             NaN             N            NaN             NaN  
4           CARGO           Y-L           PFAK           SS-SR  
..            ...           ...            ...             ...  
746         CARGO           NaN            NaN        

## Exercise 4: Working with Data in Python

Now that we've loaded our data into Python, we can utilize the data science tool Pandas to explore the data and answer a lot of interesting questions!


### Question 1: What are the types for aviation facilities in Alaska?

In this dataset, there is a variable called `Type` that records the type of a specific aviation facility. To answer this question, we would need to know what are the unique entries for this variable.

So naturally, we'll need to first extract this column from the entire dataset. The way we extract a column from a dataframe is similar to how we extract a column from a Python dictionary, except that we pass in the column (variable) name instead of key value:


In [25]:
df_facilities["Type"].unique()

array(['AIRPORT', 'SEAPLANE BASE', 'HELIPORT'], dtype=object)

### Question 2: Provide a table containing the information for all Seaplane Bases in Alaska

In this dataset, there is a variable called `Type` that records what is the type of a specific aviation facility. We want to filter out "Seaplane Base" on variable `Type`. Pandas provides a convenient way to perform element-wise comparison:


In [28]:
df_facilities["Type"] == "SEAPLANE_BASE"

0      False
1      False
2      False
3      False
4      False
       ...  
746    False
747    False
748    False
749    False
750    False
Name: Type, Length: 751, dtype: bool

As shown in the result above, the element-wise comparison result is returned in format of a Pandas series, with `True`s and `False`s representing if each case (row) satisfies the condition.

To perform filtering, we can pass in this filtering condition into the original dataframe:


In [30]:
df_facilities[df_facilities["Type"] == "SEAPLANE BASE"]

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
3     50017.1*C  SEAPLANE BASE       'KKI     3/30/2017    AAL           NONE   
5      50022.*C  SEAPLANE BASE       'KQA     3/30/2017    AAL           NONE   
8    50027.52*C  SEAPLANE BASE       'Z33     3/30/2017    AAL           NONE   
15    50031.5*C  SEAPLANE BASE      'AK81     3/30/2017    AAL           NONE   
24    50035.2*C  SEAPLANE BASE       '3C3     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
725  50877.32*C  SEAPLANE BASE       'MFN     3/30/2017    AAL           NONE   
734  50877.01*C  SEAPLANE BASE       '2X2     3/30/2017    AAL           NONE   
737    50905.*C  SEAPLANE BASE       '68A     3/30/2017    AAL           NONE   
746  50920.12*C  SEAPLANE BASE       '2Y3     3/30/2017    AAL           NONE   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   

    State StateName               County CountyState  ...  \
3      AK    ALASKA               BETHEL          AK  ...   
5      AK    ALASKA       ALEUTIANS EAST          AK  ...   
8      AK    ALASKA          BRISTOL BAY          AK  ...   
15     AK    ALASKA        KODIAK ISLAND          AK  ...   
24     AK    ALASKA            ANCHORAGE          AK  ...   
..    ...       ...                  ...         ...  ...   
725    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
734    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
737    AK    ALASKA  WRANGELL-PETERSBURG          AK  ...   
746    AK    ALASKA      SKAGWAY-YAKUTAT          AK  ...   
749    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   

    AirportPositionSource AirportPositionSourceDate AirportElevationSource  \
3                     NaN                       NaN                    NaN   
5                   STATE                  00:00:00                  STATE   
8                     NaN                       NaN                    NaN   
15                    NaN                       NaN                    NaN   
24                  OWNER                  00:00:00                  OWNER   
..                    ...                       ...                    ...   
725                 OWNER                  00:00:00                  OWNER   
734               FAA-EST                  00:00:00                  STATE   
737                   NaN                       NaN                    NaN   
746               FAA-EST                  00:00:00                    NaN   
749                   NaN                       NaN                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
3                          NaN                   NaN              NaN   
5                     00:00:00                   NaN              NaN   
8                          NaN                   NaN              TIE   
15                         NaN                   NaN              NaN   
24                    00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
725                   00:00:00                   NaN              NaN   
734                   00:00:00                   NaN              TIE   
737                        NaN                   NaN              NaN   
746                        NaN                   NaN              TIE   
749                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
3             NaN             N            NaN             NaN  
5             NaN             Y            NaN             NaN  
8             NaN           NaN            NaN             NaN  
15            NaN             N            NaN             NaN  
24            NaN             Y            NaN             NaN  
..            ...           ...            ...             ...  
725           NaN           Na

df_facilities[df_facilities["Type"] == "SEAPLANE BASE"]` can be interpreted as:

> Filter dataframe `df_facilties` conditioning on variable `Type` being strictly equal to `"SEAPLANE BASE"`.

And the table shown above is the answer to Question 2.


### Question 3: What are the 5 highest aviation facilities in Alaska?

In this dataset, the `ARPElevation` variable records the elevation of aviation facilities in Alaska in feet. What we will do is to sort the rows of this dataset in reverse based on the value of column `ARPElevation`. This can be done with `sort_values` method of Pandas dataframe:


In [32]:
df_facilities.sort_values(by = "ARPElevation", ascending = False)

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
584   50739.1*A        AIRPORT      'AA09     3/30/2017    AAL           NONE   
630  50738.20*A        AIRPORT      '93AK     3/30/2017    AAL           NONE   
265   50324.6*A        AIRPORT       '4Z5     3/30/2017    AAL           NONE   
582  50738.01*A        AIRPORT      '2AK7     3/30/2017    AAL           NONE   
353  50439.18*A        AIRPORT      'AK80     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   
261    50321.*C  SEAPLANE BASE       'OOH     3/30/2017    AAL           NONE   
289  50385.06*H       HELIPORT      '73AK     3/30/2017    AAL           NONE   
714   50873.3*C  SEAPLANE BASE       'KWP     3/30/2017    AAL           NONE   
713    50872.*C  SEAPLANE BASE       'KWF     3/30/2017    AAL           NONE   

    State StateName               County CountyState  ...  \
584    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
630    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
265    AK    ALASKA  SOUTHEAST FAIRBANKS          AK  ...   
582    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
353    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
..    ...       ...                  ...         ...  ...   
749    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   
261    AK    ALASKA      SKAGWAY-YAKUTAT          AK  ...   
289    AK    ALASKA     ALEUTIAN ISLANDS          AK  ...   
714    AK    ALASKA        KODIAK ISLAND          AK  ...   
713    AK    ALASKA      PRINCE OF WALES          AK  ...   

    AirportPositionSource AirportPositionSourceDate AirportElevationSource  \
584                   NaN                       NaN                    NaN   
630                 OWNER                  00:00:00                  OWNER   
265            FAA OE/AAA                  00:00:00             FAA OE/AAA   
582                   NaN                       NaN                    NaN   
353                 OWNER                  00:00:00                  OWNER   
..                    ...                       ...                    ...   
749                   NaN                       NaN                    NaN   
261                   NaN                       NaN                    NaN   
289                 OWNER                  00:00:00                  OWNER   
714                   NaN                       NaN                    NaN   
713                   NaN                       NaN                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
584                        NaN                   NaN              NaN   
630                   00:00:00                   NaN              NaN   
265                   00:00:00                   NaN              NaN   
582                        NaN                   NaN              NaN   
353                   00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
749                        NaN                   NaN              NaN   
261                        NaN                   NaN              TIE   
289                   00:00:00                   NaN              NaN   
714                        NaN                   NaN              NaN   
713                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
584           NaN             Y            NaN             NaN  
630         CARGO             Y            NaN             NaN  
265           NaN             Y            NaN             NaN  
582           NaN           NaN            NaN             NaN  
353           NaN             Y            NaN             NaN  
..            ...           ...            ...             ...  
749         CARGO             

When `ascending = True`, the values are arranged from smallest to largest. This means that with `ascending = False`, the values will be arranged from largest to smallest, allowing us to sort the aviation facilities from highest to lowest.

To sort the dataset, we need to know:

*   The name(s) of column(s) to sort by, and
*   The sorting order.

After sorting the dataset, we'll extract the first five rows which will contain the five highest aviation facilities, thus answering Question 3. This can be done with `head` method of Pandas dataframe.


In [35]:
df_facilities.sort_values(by = "ARPElevation", ascending = False).head(n = 5)

SiteNumber     Type LocationID EffectiveDate Region DistrictOffice State  \
584   50739.1*A  AIRPORT      'AA09     3/30/2017    AAL           NONE    AK   
630  50738.20*A  AIRPORT      '93AK     3/30/2017    AAL           NONE    AK   
265   50324.6*A  AIRPORT       '4Z5     3/30/2017    AAL           NONE    AK   
582  50738.01*A  AIRPORT      '2AK7     3/30/2017    AAL           NONE    AK   
353  50439.18*A  AIRPORT      'AK80     3/30/2017    AAL           NONE    AK   

    StateName               County CountyState  ... AirportPositionSource  \
584    ALASKA    MATANUSKA-SUSITNA          AK  ...                   NaN   
630    ALASKA    MATANUSKA-SUSITNA          AK  ...                 OWNER   
265    ALASKA  SOUTHEAST FAIRBANKS          AK  ...            FAA OE/AAA   
582    ALASKA    MATANUSKA-SUSITNA          AK  ...                   NaN   
353    ALASKA    MATANUSKA-SUSITNA          AK  ...                 OWNER   

    AirportPositionSourceDate AirportElevationSource  \
584                       NaN                    NaN   
630                  00:00:00                  OWNER   
265                  00:00:00             FAA OE/AAA   
582                       NaN                    NaN   
353                  00:00:00                  OWNER   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
584                        NaN                   NaN              NaN   
630                   00:00:00                   NaN              NaN   
265                   00:00:00                   NaN              NaN   
582                        NaN                   NaN              NaN   
353                   00:00:00                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
584           NaN             Y            NaN             NaN  
630         CARGO             Y            NaN             NaN  
265           NaN             Y            NaN             NaN  
582           NaN           NaN            NaN             NaN  
353           NaN             Y            NaN             NaN  

[5 rows x 103 columns]

### Question 4: What is the average elevation of aviation facilities?

In this dataset, we know that `ARPElevation` tells us the elevation of aviation facilities in Alaska. So how do we find the average or, in other words, the mean of those numbers?

Pandas has a built in function `mean()` that we can use to easily calculate the average of that column.

Let's take our dataframe and apply that function!


In [36]:
df_facilities["ARPElevation"].mean()

np.float64(425.93209054593876)

As we can see from that result, the average elevation of aviation facilities is almost 426 ft.


### Question 5: What are the highest and lowest elevation values of the aviation facilities?

How do we go about solving this question? Well, similar to the previous question, Pandas has built-in functions that help us find the maximum (largest) and minimum (lowest) values in a column. These functions are `max()` and `min()`, respectively.

#### Task A: Find the highest elevation value


In [39]:
df_facilities["ARPElevation"].max()

np.int64(3984)

#### Task B: Find the lowest elevation value

In [41]:
df_facilities["ARPElevation"].min()

np.int64(0)

The maximum and minimum values of elevation are 3984 ft and 0 ft respectively.

## Exercise 5: Write to `xlsx` Files!

After getting the results we want, we often want to save those results on our computer for future use. We can save a dataframe to our computer directly in Python!

For example, let's say we want to store the answer to Question 1 in a file called `seaplane.xlsx`.

First, let's save the result to a local variable:


In [43]:
df_seaplane = df_facilities[df_facilities["Type"] == "SEAPLANE BASE"]

Now, the answer to Question 1 is saved in the variable `df_seaplane`.

To save a dataframe in `xlsx` format, we can use the `to_excel` method. This will write the contents of the object, `df_seaplane`, into an Excel `xlsx` file.


In [45]:
df_seaplane.to_excel("./seaplane.xlsx")

We've successfully saved the table from Question 1 to a file! If we want to see the file we saved, check the file panel on the left.